fmmax/vector_fourier.py

```diff
@@ -0,0 +1,445 @@
+    primitive_lattice_vectors: basis.LatticeVectors,
+    fourier_loss_weight: float,
+) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """Compute tangent vector field using the Jones direct method."""
+    return compute_tangent_field(
+        arr=arr,
+        expansion=expansion,
+        primitive_lattice_vectors=primitive_lattice_vectors,
+        use_jones_direct=True,
+        fourier_loss_weight=fourier_loss_weight,
+    )
+
+
+def compute_field_pol(
+    arr: jnp.ndarray,
+    expansion: basis.Expansion,
+    primitive_lattice_vectors: basis.LatticeVectors,
+    fourier_loss_weight: float,
+) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """Compute tangent vector field using the Pol method."""
+    return compute_tangent_field(
+        arr=arr,
+        expansion=expansion,
+        primitive_lattice_vectors=primitive_lattice_vectors,
+        use_jones_direct=False,
+        fourier_loss_weight=fourier_loss_weight,
+    )
+
+
+def compute_field_jones(
+    arr: jnp.ndarray,
+    expansion: basis.Expansion,
+    primitive_lattice_vectors: basis.LatticeVectors,
+    fourier_loss_weight: float,
+) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """Compute tangent vector field using the Jones method."""
+    tx, ty = compute_tangent_field(
+        arr=arr,
+        expansion=expansion,
+        primitive_lattice_vectors=primitive_lattice_vectors,
+        use_jones_direct=False,
+        fourier_loss_weight=fourier_loss_weight,
+    )
+    jxjy = normalize_jones(jnp.stack([tx, ty], axis=-1))
+    jx = jxjy[..., 0]
+    jy = jxjy[..., 1]
+    return jx, jy
+
+
+def compute_field_normal(
+    arr: jnp.ndarray,
+    expansion: basis.Expansion,
+    primitive_lattice_vectors: basis.LatticeVectors,
+    fourier_loss_weight: float,
+) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """Compute tangent vector field using the Normal method."""
+    tx, ty = compute_tangent_field(
+        arr=arr,
+        expansion=expansion,
+        primitive_lattice_vectors=primitive_lattice_vectors,
+        use_jones_direct=False,
+        fourier_loss_weight=fourier_loss_weight,
+    )
+    txty = normalize_elementwise(jnp.stack([tx, ty], axis=-1))
+    tx = txty[..., 0]
+    ty = txty[..., 1]
+    return tx, ty
+
+
+# -----------------------------------------------------------------------------
+# Underlying functions used to compute all variants of the tangent fields.
+# -----------------------------------------------------------------------------
+
+
+def compute_tangent_field(
+    arr: jnp.ndarray,
+    expansion: basis.Expansion,
+    primitive_lattice_vectors: basis.LatticeVectors,
+    use_jones_direct: bool,
+    fourier_loss_weight: float,
+    steps: int = 1,
+) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    """Compute the tangent vector field for `arr`.
+
+    The calculation finds the minimum of a quadratic loss function using a single
+    Newton iteration. Rather than optimizing the real-space tangent field, the
+    Fourier coefficients are directly optimized.
```

Comment:
This is the main innovation to these approaches, right?

Reply:
Essentially, yes. More explicitly, the change is is a (1) a new loss function, (2) a new parameterization (solving for the Fourier coefficients directly, rather than the vector field on the real-space grid), and (3) the use of Newton method rather than a first order method to minimize the loss. The update to fix vector fields for odd-shaped unit cells is a further change.